Dev r34 config query 2

sidey79 opened the pull request:
--- Nur eine Idee, nicht getestet ----


* **Please check if the PR fulfills these requirements**
<!-- Please do not remove these checkboxes. Check them if you have done the action behind this point -->
- [ ] Tests for the changes have been added / modified (needed for for bug fixes / features)
- [ ] commandref has been added / updated (needed for bug fixes / features)
- [ ] CHANGED has been updated (needed for bug fixes / features)


<!--Please specify if this is a bugfix, feature or update of docs and remove the other options -->

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)



* **What is the current behavior?** (You can also link to an open issue here)


* **What is the new behavior (if this is a feature change)?**



* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)



* **Other information**:

sidey79 committed:
00_SIGNALduino.pm

Idea to make multiple get commands more easy and robust to handle

actions-user committed:
Automaitc updated controls and CHANGED

sidey79 commented:
@homeautouser
@elektrin-bbs

Die Idee der Änderung ist, dass die Get Routine prüft ob noch auf eine Antwort gewartet wird und dann verzögert.

Die Abfrage / erneuerung der cc1101 Parameter basiert auf Änderungen an Registern und nicht auf Anpassungen rein über Get Befehle.

Ich habe den Code allerdings nicht getestet.
Es ist eher eine Konzeptidee.

HomeAutoUser commented:
Ich werde mir ihn ansehen und testen. Es wird sicherlich erst morgen.
Das eine Prüfung stattfindet ob noch etwas erwartet wird klingt erstmal gut.

sidey79 committed:
syntax fixes

sidey79 committed:
Merge branch 'dev-r34_config_query_2' of github.com:RFD-FHEM/RFFHEM into dev-r34_config_query_2

actions-user committed:
Automaitc updated controls and CHANGED

sidey79 committed:
tests/test_sub_SIGNALduino_Get-definition.txt

clear uCcmd after every subtest

sidey79 committed:
Merge branch 'dev-r34_config_query_2' of github.com:RFD-FHEM/RFFHEM into dev-r34_config_query_2

sidey79 committed:
tests/test_DeviceData_rmsg-definition.txt

fix restore of attribute whitelist_IDs after subtests

sidey79 committed:
tests/test_sub_SIGNALduino_Parse_MU-definition.txt

fix restore of attribute whitelist_IDs after subtests

HomeAutoUser commented:
@sidey79 
ich habe soeben mal deinen Branch getestet. Der Ansatz ist gut.
Es besteht noch Handlungsbedarf bevor ich das "durchgehen" lassen würde ;-)

- Automatisches abrufen der `cc1101_config` und `cc1101_config_ext` funktioniert sobald man dort etwas verstellt.

```
		if ($hash->{version} =~ m/cc1101/) {
			$hash->{cc1101_available} = 1;
			SIGNALduino_Get($hash, $name,"ccconf");
			SIGNALduino_Get($hash, $name,"ccpatable");
		}
```
Hiervon wird nur das `ccconf` ausgeführt. `ccpatable` keine Wirkung.

1) abrufen von `cc1101_patable` beim verstellen geht nicht
2) beim Neustart von FHEM oder beim RESET vom cc1101 werden nicht die aktuellen `cc1101_config` und `cc1101_config_ext` bzw. `cc1101_patable`  abgerufen.
3) PERL WARNING beim shutdown restart
`2020.03.09 16:08:34 1: PERL WARNING: Use of uninitialized value in addition (+) at ./FHEM/00_SIGNALduino.pm line 1035.`

HomeAutoUser commented on FHEM/00_SIGNALduino.pm:
```diff
@@ -1343,8 +1349,8 @@ sub SIGNALduino_CheckVersionResp
 			$hash->{cc1101_available} = 1;
-			InternalTimer(gettimeofday() + SDUINO_GET_CONFIGQUERY_DELAY, "SIGNALduino_delayed_Get", "SIGNALduino_delayed_Get:$name:ccconf", 0);
-			InternalTimer(gettimeofday() + SDUINO_GET_CONFIGQUERY_DELAY + 1, "SIGNALduino_delayed_Get", "SIGNALduino_delayed_Get:$name:ccpatable", 0);
+			SIGNALduino_Get($hash, $name,"ccconf");
+			SIGNALduino_Get($hash, $name,"ccpatable");
```
Hier wird nur jeweils eines ausgeführt.
Wenn ich `ccconf` auskommentiere, so erhalte ich einen Wert von `ccpatable`. Beides zusammen einkommentiert, so funktioniert nur eines.

HomeAutoUser commented on FHEM/00_SIGNALduino.pm:
```diff
@@ -1028,13 +1029,18 @@ sub SIGNALduino_GetResponseUpdateReading
+	my(undef,$name,@cmds) = split(':', shift);
+	my $hash = $defs{$name};
+  
+	if (exists($hash->{ucCmd}) && $hash->{ucCmd}{timenow}+10 > time() ) {
```
Beim ersten Durchlauf nach einem `shutdown restart` ist  `$hash->{ucCmd}{timenow}` nicht existent und somit kann die Addition von 10 nicht erfolgen. Das wird auch die Perl Warning hervorrufen.

Eine Suche nach `$hash->{ucCmd}{timenow}` über die komplette 00_SIGNALduino.pm zeigt auch, das dieser Hashwert nicht gefüllt wird.

sidey79 replied:
Kannst Du die Stellen benennen an denen $hash->{ucCmd}{timenow} nicht aber $hash->{ucCmd} gesetzt sind?

HomeAutoUser replied:
Meine Antwort bezog sich auf einen Stand bevor du die Änderungen vollzogen hattest. Nach dem jetzigen kontrollieren habe ich keine Stelle gefunden wo $hash->{ucCmd}{timenow} fehlt.

HomeAutoUser commented on FHEM/00_SIGNALduino.pm:
```diff
@@ -1028,13 +1029,18 @@ sub SIGNALduino_GetResponseUpdateReading
+  
+	if (exists($hash->{ucCmd}) && $hash->{ucCmd}{timenow}+10 > time() ) {
+		$hash->{logMethod}->($hash->{NAME}, 5, "$name: SIGNALduino_delayed_Get, ".split(":",@cmds)." delayed");
+		main::InternalTimer(main::gettimeofday() + main::SDUINO_GET_CONFIGQUERY_DELAY, "SIGNALduino_delayed_Get", "SIGNALduino_delayed_Get:$name:".split(":",@cmds), 0);
```
Mit einem `split(":",@cmds)` erhält du nur die Anzahl der darin stehenden Cmds.
`2020.03.09 16:25:55 5: nano_433Mhz: SIGNALduino_delayed_Get, 1 executed`

Wir benötigen dort die Getbefehle (Bsp: ccpatable) als Wert. Man müsste vielleicht dort `$cmds[0]` nutzen wenn das ausgeführte Cmd auch aus der Liste wieder entfernt wird um an die nächsten Cmds zu kommen.

sidey79 replied:
okay join ist der Korrekte Befehl. Split ist ja quatsch

sidey79 commented:
@homeautouser

Danke für das Feedback.

Die Fehler sollte man noch abwenden, das sehe ich auch so. :)

sidey79 committed:
tests\test_sub_SIGNALduino_CheckVersionResp-definition.txt

extended test to cover register query after valid version is rereived

HomeAutoUser committed:
Update 00_SIGNALduino.pm

- rename sub and revised logoutput
- fix some parameter https://github.com/HomeAutoUser/RFFHEM/pull/9#issuecomment-596668609

actions-user committed:
Automaitc updated controls and CHANGED

HomeAutoUser commented:
@sidey79 
mit https://github.com/RFD-FHEM/RFFHEM/commit/c51f11989e769237d4c0a70c99cec63f6e8f5d52 habe ich die Dinge soweit angepasst bis auf den ersten Abruf beim shutdown restart. Da habe ich das Gefühl, das dort etwas überschrieben wird und nur der `ccpatable` Befehl ausgeführt wird.

Beweis:

```
2020.03.10 19:13:40 2: SIGNALduino_Board: CheckVersionResp, initialized v3.4.2_dev_02.03
2020.03.10 19:13:40 5: SIGNALduino_Board: SimpleWrite, XE
2020.03.10 19:13:40 3: SIGNALduino_Board: CheckVersionResp, enable receiver (XE) 
2020.03.10 19:13:40 1: PERL WARNING: Use of uninitialized value $text in concatenation (.) or string at fhem.pl line 989.
2020.03.10 19:13:40 5: 
2020.03.10 19:13:40 5: SIGNALduino_Board: Get ccconf executed
2020.03.10 19:13:40 5: SIGNALduino_Board: AddSendQueue, SIGNALduino_Board: C0DnF (1)
2020.03.10 19:13:40 5: SIGNALduino_Board: Get_delayed, ccpatable delayed
2020.03.10 19:13:40 5: SIGNALduino_Board: Read, try asyncOutput of message V 3.4.0-dev SIGNALduino cc1101 (chip CC1101) - compiled at Mar 10 2020 18:37:19
2020.03.10 19:13:40 4: SIGNALduino_Board: HandleWriteQueue, called
2020.03.10 19:13:40 4: SIGNALduino_Board: SendFromQueue, called
2020.03.10 19:13:40 5: SIGNALduino_Board: SimpleWrite, C0DnF
2020.03.10 19:13:40 4: SIGNALduino_Board: Read, msg: C0Dn11=10B071F7C43023B900070018146C070091
2020.03.10 19:13:40 5: SIGNALduino_Board: Parse, noMsg: C0Dn11=10B071F7C43023B900070018146C070091
2020.03.10 19:13:40 4: SIGNALduino_Board: HandleWriteQueue, called
2020.03.10 19:13:40 4: SIGNALduino_Board: HandleWriteQueue, nothing to send, stopping timer
2020.03.10 19:13:40 5: 
2020.03.10 19:13:40 5: SIGNALduino_Board: Get ccpatable executed
2020.03.10 19:13:40 5: SIGNALduino_Board: AddSendQueue, SIGNALduino_Board: C3E (1)
2020.03.10 19:13:40 4: SIGNALduino_Board: HandleWriteQueue, called
2020.03.10 19:13:40 4: SIGNALduino_Board: SendFromQueue, called
2020.03.10 19:13:40 5: SIGNALduino_Board: SimpleWrite, C3E
2020.03.10 19:13:40 4: SIGNALduino_Board: Read, msg: C3E = 00 12 00 00 00 00 00 00
2020.03.10 19:13:40 5: SIGNALduino_Board: Parse, noMsg: C3E = 00 12 00 00 00 00 00 00
2020.03.10 19:13:40 5: SIGNALduino_Board: Read, msg: regexp=^C3E\s=\s.* cmd=ccpatable msg=C3E = 00 12 00 00 00 00 00 00
2020.03.10 19:13:40 3: SIGNALduino_Board: CheckCcpatableResponse, patable: 12
2020.03.10 19:13:40 5: SIGNALduino_Board: CheckCcpatableResponse, patable: -30_dBm
2020.03.10 19:13:40 5: SIGNALduino_Board: Read, try asyncOutput of message C3E = 00 12 00 00 00 00 00 00 => -30_dBm
```

Ein richtiger Ablauf von dem Befehl ccconf  müsste so aussehen:

```
2020.03.10 18:57:05 5: SIGNALduino_Board: Get ccconf executed
2020.03.10 18:57:05 5: SIGNALduino_Board: AddSendQueue, SIGNALduino_Board: C0DnF (1)
2020.03.10 18:57:05 4: SIGNALduino_Board: HandleWriteQueue, called
2020.03.10 18:57:05 4: SIGNALduino_Board: SendFromQueue, called
2020.03.10 18:57:05 5: SIGNALduino_Board: SimpleWrite, C0DnF
2020.03.10 18:57:05 4: SIGNALduino_Board: Read, msg: C0Dn11=10B07157C43023B900070018146C070091
2020.03.10 18:57:05 5: SIGNALduino_Board: Parse, noMsg: C0Dn11=10B07157C43023B900070018146C070091
2020.03.10 18:57:05 5: SIGNALduino_Board: Read, msg: regexp=C0Dn11=[A-F0-9a-f]+ cmd=ccconf msg=C0Dn11=10B07157C43023B900070018146C070091
2020.03.10 18:57:05 5: SIGNALduino_Board: Read, try asyncOutput of message Freq: 433.920 MHz, Bandwidth: 325 KHz, rAmpl: 42 dB, sens: 8 dB, DataRate: 5603.79 Baud, Modulation: ASK/OOK, Syncmod: No preamble/sync
```

Ich bin mir aber nicht sicher ob es ein Zeitfaktor ist, weil anfangs schon etwas läuft oder ob es ein Verarbeitungsfehler in Form einer Überscheibung ist.

@elektron-bbs kannst du bitte mal parallel mit testen wie es sich verhält.

EDIT: Zusätzlich ist mir aufgefallen @sidey79, das beim Abrufen der Version ständig ccconf und ccpatable abgerufen werden. Ist das Absicht ???

sidey79 committed:
00_SIGNALduino.pm

define $hash->{ucCmd}->{timenow} if ucCmd is specified on serval places

Updated test_sub_SIGNALduino_CheckVersionResp-definition.txt

sidey79 committed:
Merge branch 'dev-r34_config_query_2' of github.com:RFD-FHEM/RFFHEM into dev-r34_config_query_2

actions-user committed:
Automaitc updated controls and CHANGED

sidey79 commented:
> Ich bin mir aber nicht sicher ob es ein Zeitfaktor ist, weil anfangs schon etwas läuft oder ob es ein Verarbeitungsfehler in Form einer Überscheibung ist.

Es lag einfach daran, dass timenow nicht gesetzt war und dann wird mit der Anpassung der Befehl überschrieben


> EDIT: Zusätzlich ist mir aufgefallen @sidey79, das beim Abrufen der Version ständig ccconf und ccpatable abgerufen werden. Ist das Absicht ???

Ich hatte das von dir übernommen und dachte, das wäre so, damit nach dem Initialisieren des CC1101 auch gleich die Readings mit erzeugt werden.
Das passt vielleicht besser in SIGNALduino_DoInit

sidey79 committed:
00_SIGNALduino.pm

try to query ccconf and ccpatable only if device is initalized

actions-user committed:
Automaitc updated controls and CHANGED

sidey79 committed:
00_SIGNALduino.pm

log if timenow is not set
corrected config requests per uC command

actions-user committed:
Automaitc updated controls and CHANGED

sidey79 committed:
tests/test_sub_SIGNALduino_Get_Callback-definition.txt

 cleanup after test

sidey79 committed:
tests/test_get_availableFirmware-definition.txt

Cleanup getCmd before testing

sidey79 committed:
Merge branch 'dev-r34_config_query_2' of github.com:RFD-FHEM/RFFHEM into dev-r34_config_query_2

sidey79 committed:
00_SIGNALduino.pm

SIGNALduino_Get_delayed support more than one arg for cmds

actions-user committed:
Automaitc updated controls and CHANGED

HomeAutoUser commented:
> 
> 
> Mist, das könnte gut möglich sein.
> 
> Irgendeine Idee, wie wir das machen, dass es nur beim Neuinitalisieren abgefragt wird?

Bitte mal https://github.com/RFD-FHEM/RFFHEM/tree/dev-r34_config_query_2 testen. Das sollte den Fall erschlagen

HomeAutoUser committed:
Update 00_SIGNALduino.pm

- revised logoutput -> $hash->{ucCmd}output always HASH, now ... Get_delayed, timenow was missing, set "value"
- added option ccconf & ccconf if RESTART and programm shutdown restart

actions-user committed:
Automaitc updated controls and CHANGED

sidey79 commented:
So, die ganze Anpassung ist mit Tests aktuell nicht abgedeckt, aber das packe ich heute nicht mehr

HomeAutoUser commented:
>So, die ganze Anpassung ist mit Tests aktuell nicht abgedeckt, aber das packe ich heute nicht mehr

Das könnten wir ja noch Anpassen bzw. du kannst die Fälle abdecken.

> ... @sidey79  wants to merge 26 commits into HomeAutoUser:dev-r34_config_query from RFD-FHEM:dev-r34_config_query_2

Wenn ich das richtig sehe, so wird nach der Übernahme des PR´s die Daten ja automatisch in den PR https://github.com/RFD-FHEM/RFFHEM/pull/800 überführt welche ja dennoch "offen" bleibt.
So würde ich hier übernehmen um dann den offenen PR#800 zum Schluss mit allen anpassungen zu vollenden.
